File: archive/2024-06-01/tree/T5_levelOrder.java
package org.review2024summer.tree;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;

public class T5_levelOrder {
    public List<List<Integer>> levelOrderByIter(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        if (root != null)
            queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new ArrayList<>();
            while (size-- > 0) {
                TreeNode curr = queue.poll();
                level.add(curr.val);
                if (curr.left != null)
                    queue.add(curr.left);
                if (curr.right != null)
                    queue.add(curr.right);
            }
            res.add(level);
        }
        return res;
    }

    public List<List<Integer>> levelOrderByRes(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        orderByRes(root, 0, res);
        return res;
    }

    public void orderByRes(TreeNode root, int depth, List<List<Integer>> res) {
        if (root == null)
            return;
        if (res.size() == depth)
            res.add(new ArrayList<>());
        res.get(depth).add(root.val);
        orderByRes(root.left, depth + 1, res);
        orderByRes(root.right, depth + 1, res);
    }

    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        if (root != null)
            queue.add(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            ArrayList<Integer> level = new ArrayList<>();
            while (size-- > 0) {
                TreeNode curr = queue.poll();
                level.add(curr.val);
                if (curr.left != null) queue.add(curr.left);
                if (curr.right != null) queue.add(curr.right);
            }
            res.add(0, level);
        }
        return res;
    }

    public List<Integer> rightSideView(TreeNode root) {
        ArrayList<Integer> res = new ArrayList<>();
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        if (root != null)
            queue.push(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode curr = queue.poll();
                if (i == size - 1)
                    res.add(curr.val);
                if (curr.left != null) queue.add(curr.left);
                if (curr.right != null) queue.add(curr.right);
            }
        }
        return res;
    }

    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> res = new ArrayList<>();
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        if (root != null)
            queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            double sum = 0;
            for (int i = 0; i < size; i++) {
                TreeNode curr = queue.poll();
                sum += curr.val;
                if (curr.left != null)
                    queue.add(curr.left);
                if (curr.right != null)
                    queue.add(curr.right);
            }
            res.add(sum / size);
        }
        return res;
    }

    public List<List<Integer>> levelOrder(NodeWithChild root) {
        List<List<Integer>> res = new ArrayList<>();
        ArrayDeque<NodeWithChild> queue = new ArrayDeque<>();
        if (root != null)
            queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new ArrayList<>();
            while (size-- > 0) {
                NodeWithChild curr = queue.poll();
                level.add(curr.val);
                List<NodeWithChild> children = curr.children;
                queue.addAll(children);
            }
            res.add(level);
        }
        return res;
    }

    public List<Integer> largestValues(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        if (root != null)
            queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            int max = Integer.MIN_VALUE;
            while (size-- > 0) {
                TreeNode curr = queue.poll();
                max = Math.max(max, curr.val);
                if (curr.left != null)
                    queue.add(curr.left);
                if (curr.right != null)
                    queue.add(curr.right);
            }
            res.add(max);
        }
        return res;
    }

    public NodeWithNext connect(NodeWithNext root) {
        ArrayDeque<NodeWithNext> queue = new ArrayDeque<>();
        if (root != null)
            queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            ArrayDeque<NodeWithNext> level = new ArrayDeque<>();
            while (size-- > 0) {
                NodeWithNext curr = queue.poll();
                level.add(curr);
                if (curr.left != null)
                    queue.add(curr.left);
                if (curr.right != null)
                    queue.add(curr.right);
            }

            NodeWithNext curr = level.pollFirst();
            while (curr != null) {
                NodeWithNext tmp = level.pollFirst();
                curr.next = tmp;
                curr = tmp;
            }
        }
        return root;
    }

    public int maxDepth(TreeNode root) {
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        int depth = 0;
        if (root != null)
            queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            depth++;
            while (size-- > 0) {
                TreeNode curr = queue.poll();
                if (curr.left != null)
                    queue.add(curr.left);
                if (curr.right != null)
                    queue.add(curr.right);
            }
        }
        return depth;
    }

    public int minDepth(TreeNode root) {
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        int depth = 0;
        if (root != null)
            queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            depth++;
            while (size-- > 0) {
                TreeNode curr = queue.poll();
                if (curr.left == null && curr.right == null)
                    return depth;
                if (curr.left != null)
                    queue.add(curr.left);
                if (curr.right != null)
                    queue.add(curr.right);
            }
        }
        return depth;
    }



































}

File: templates/python/code-carl/ch06_tree/t5_level_order.py
"""LeetCode Problems:
- 102. Binary Tree Level Order Traversal
  https://leetcode.cn/problems/binary-tree-level-order-traversal/
- 107. Binary Tree Level Order Traversal II
  https://leetcode.cn/problems/binary-tree-level-order-traversal-ii/
- 199. Binary Tree Right Side View
  https://leetcode.cn/problems/binary-tree-right-side-view/
- 637. Average of Levels in Binary Tree
  https://leetcode.cn/problems/average-of-levels-in-binary-tree/
- 429. N-ary Tree Level Order Traversal
  https://leetcode.cn/problems/n-ary-tree-level-order-traversal/
- 515. Find Largest Value in Each Tree Row
  https://leetcode.cn/problems/find-largest-value-in-each-tree-row/
- 116. Populating Next Right Pointers in Each Node
  https://leetcode.cn/problems/populating-next-right-pointers-in-each-node/
- 104. Maximum Depth of Binary Tree
  https://leetcode.cn/problems/maximum-depth-of-binary-tree/
- 111. Minimum Depth of Binary Tree
  https://leetcode.cn/problems/minimum-depth-of-binary-tree/
"""

from typing import List, Optional

from tree_node import Node, TreeNode


class Solution:
    def levelOrderByIter(self, root: Optional[TreeNode]) -> List[List[int]]:
        pass

    def levelOrderByRes(self, root: Optional[TreeNode]) -> List[List[int]]:
        pass

    def orderByRes(self, root: Optional[TreeNode], depth: int, res: List[int]) -> None:
        pass

    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        pass

    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        pass

    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        pass

    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def connect(self, root: Optional[Node]) -> Optional[Node]:
        pass

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        pass

    def minDepth(self, root: Optional[TreeNode]) -> int:
        pass


if __name__ == "__main__":
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))

    solution = Solution()
    assert solution.levelOrderByIter(root) == [[3], [9, 20], [15, 7]]
    solution = Solution()
    assert solution.levelOrderByRes(root) == [[3], [9, 20], [15, 7]]
    solution = Solution()
    assert solution.levelOrderBottom(root) == [[15, 7], [9, 20], [3]]
    solution = Solution()
    assert solution.rightSideView(root) == [3, 20, 7]
    solution = Solution()
    assert solution.averageOfLevels(root) == [3.0, 14.5, 11.0]
    solution = Solution()
    assert solution.largestValues(root) == [3, 20, 15]
    solution = Solution()
    assert solution.maxDepth(root) == 3
    solution = Solution()
    assert solution.minDepth(root) == 2

    nroot = Node(1, Node(2, Node(4), Node(5)), Node(3, Node(6), Node(7)))
    solution = Solution()
    connected = solution.connect(nroot)
    assert connected.left.next is connected.right
    assert connected.left.left.next is connected.left.right
    assert connected.left.right.next is connected.right.left

    print("All tests passed.")
